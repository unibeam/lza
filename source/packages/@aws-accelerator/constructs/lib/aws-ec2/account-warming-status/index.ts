/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
/**
 * aws-ec2-pre-warm-account-status - lambda handler
 *
 * @param event
 * @returns
 */

import { SSMClient, PutParameterCommand } from '@aws-sdk/client-ssm';
import {
  EC2Client,
  DescribeInstancesCommand,
  DescribeVpcsCommand,
  DescribeSubnetsCommand,
  DeleteVpcCommand,
  DeleteSubnetCommand,
  TerminateInstancesCommand,
} from '@aws-sdk/client-ec2';
import { AdaptiveRetryStrategy } from '@aws-sdk/util-retry';
import { delay } from '@aws-accelerator/utils/lib/throttle';

const solutionId = process.env['SOLUTION_ID'] ?? '';
const retryStrategy = new AdaptiveRetryStrategy(() => Promise.resolve(5));
const ec2Client = new EC2Client({ customUserAgent: solutionId, retryStrategy });
const ssmClient = new SSMClient({ customUserAgent: solutionId, retryStrategy });

type InstanceDetails = {
  instanceId: string | undefined;
  minutesSinceLaunch: number | undefined;
};

//eslint-disable-next-line @typescript-eslint/no-explicit-any
export async function handler(event: any): Promise<
  | {
      IsComplete: boolean;
    }
  | undefined
> {
  console.log(event);
  const ssmPrefix: string = event.ResourceProperties['ssmPrefix'];
  const instanceDetails = await getInstanceDetails();

  // if no instance was found, it has been deleted return
  if (instanceDetails.instanceId === undefined) {
    return { IsComplete: true };
  }

  if (instanceDetails.minutesSinceLaunch! > 15) {
    console.log('Account warming time reached');
    await terminateInstance(instanceDetails.instanceId);
    await deleteVpc();
    await setSSMParameter(ssmPrefix, 'true');
    return {
      IsComplete: true,
    };
  }
  console.log('Waiting for warmup to complete');
  return { IsComplete: false };
}

async function getInstanceDetails(): Promise<InstanceDetails> {
  console.log('Getting Instance Id');
  const instanceDetails: InstanceDetails = { instanceId: undefined, minutesSinceLaunch: undefined };
  const ec2Instances = await ec2Client.send(
    new DescribeInstancesCommand({ Filters: [{ Name: 'tag:Name', Values: ['accelerator-warm'] }] }),
  );

  if (ec2Instances.Reservations && ec2Instances.Reservations.length < 1) {
    return instanceDetails;
  }

  if (ec2Instances.Reservations![0].Instances?.length ?? 0 > 0) {
    instanceDetails.instanceId = ec2Instances.Reservations![0].Instances![0].InstanceId;
    const launchTime = ec2Instances.Reservations![0].Instances![0].LaunchTime;

    const currentTime = new Date();
    const msElapsed = currentTime.getTime() - launchTime!.getTime();
    instanceDetails.minutesSinceLaunch = msElapsed / 1000 / 60;
  }

  return instanceDetails;
}

async function deleteVpc() {
  console.log('Deleting VPC');
  const vpcId = await getVpcId();

  if (vpcId) {
    console.log('Deleting subnets');
    const subnets = await ec2Client.send(
      new DescribeSubnetsCommand({ Filters: [{ Name: 'vpc-id', Values: [vpcId] }] }),
    );
    for (const subnet of subnets.Subnets) {
      await ec2Client.send(new DeleteSubnetCommand({ SubnetId: subnet.SubnetId }));
    }
    console.log(`Deleting VPC with id: ${vpcId}`);
    await ec2Client.send(new DeleteVpcCommand({ VpcId: vpcId }));
  }
}

async function getVpcId(): Promise<string | undefined> {
  console.log('Getting VPC Id');
  const vpcs = await ec2Client.send(
    new DescribeVpcsCommand({ Filters: [{ Name: 'tag:Name', Values: ['accelerator-warm'] }] }),
  );

  if (vpcs.Vpcs?.length ?? 0 > 0) {
    return vpcs.Vpcs![0].VpcId;
  }
  return undefined;
}

async function setSSMParameter(ssmPrefix: string, parameterValue: string) {
  console.log('Updating SSM Parameter');
  try {
    await ssmClient.send(
      new PutParameterCommand({
        Name: `${ssmPrefix}/account/pre-warmed`,
        Value: parameterValue,
        Overwrite: true,
      }),
    );
  } catch (e) {
    console.log(e);
  }
}

async function terminateInstance(instanceId: string): Promise<void> {
  console.log(`Terminating EC2 instance: ${instanceId}`);
  await ec2Client.send(new TerminateInstancesCommand({ InstanceIds: [instanceId] }));
  await waitForTermination(instanceId);
}

async function waitForTermination(instanceId: string): Promise<void> {
  console.log(`Waiting for termination of instanceId: ${instanceId}`);
  let ec2Terminated = false;
  while (!ec2Terminated) {
    const statusResponse = await ec2Client.send(new DescribeInstancesCommand({ InstanceIds: [instanceId] }));
    if (
      statusResponse.Reservations[0].Instances[0].State?.Name !== 'terminated' ||
      statusResponse.Reservations[0].Instances[0].State?.Code !== 48
    ) {
      delay(15000);
      continue;
    }
    ec2Terminated = true;
  }
}
