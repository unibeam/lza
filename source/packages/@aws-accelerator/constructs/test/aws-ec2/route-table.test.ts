/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { RouteTable } from '../../lib/aws-ec2/route-table';
import { NatGateway, Vpc, Subnet } from '../../lib/aws-ec2/vpc';
import { snapShotTest } from '../snapshot-test';
import { describe, it, expect } from '@jest/globals';
import { TransitGatewayAttachment } from '../../lib/aws-ec2/transit-gateway';

const testNamePrefix = 'Construct(RouteTable): ';

//Initialize stack for snapshot test and resource configuration test
const stack = new cdk.Stack();
const key = new cdk.aws_kms.Key(stack, 'testKey');
const vpc = new Vpc(stack, 'TestVpc', {
  name: 'Test',
  ipv4CidrBlock: '10.0.0.0/16',
  internetGateway: true,
  enableDnsHostnames: false,
  enableDnsSupport: true,
  instanceTenancy: 'default',
  virtualPrivateGateway: {
    asn: 65000,
  },
});
vpc.addEgressOnlyIgw();

const rt = new RouteTable(stack, 'RouteTable', {
  name: 'TestRouteTable',
  vpc: vpc,
  tags: [{ key: 'Test-Key', value: 'Test-Value' }],
});

const subnet = new Subnet(stack, 'test-subnet', {
  name: 'test-subnet',
  routeTable: rt,
  vpc,
  availabilityZone: 'a',
  availabilityZoneId: undefined,
  ipv4CidrBlock: '10.0.2.0/24',
});

const ngw = new NatGateway(stack, 'ngw', { name: 'ngw', subnet });

const tgwAttachment = new TransitGatewayAttachment(stack, 'tgwAttachment', {
  name: 'someTgwAttachment',
  partition: 'partition',
  subnetIds: ['subnet-123'],
  transitGatewayId: 'tgw-12324',
  vpcId: vpc.vpcId,
});

rt.addTransitGatewayRoute('tg-route', 'tgw-1234', tgwAttachment, undefined, 'pl-1234', undefined, key, 10);
rt.addTransitGatewayRoute('tg', 'tg-1234', tgwAttachment, '10.0.5.0/24', undefined, undefined, key, 10);
rt.addTransitGatewayRoute('tg-ipv6', 'tg-1234', tgwAttachment, undefined, undefined, '::1', key, 10);
rt.addNatGatewayRoute('testNgwRoute', ngw.natGatewayId, '10.0.3.0/24', undefined, undefined, key, 10);
rt.addNatGatewayRoute('test2NgwRoute', ngw.natGatewayId, undefined, 'pl-1234', undefined, key, 10);
rt.addNatGatewayRoute('test3NgwRoute', ngw.natGatewayId, undefined, undefined, '::1', key, 10);
rt.addInternetGatewayRoute('testIgwRoute', '0.0.0.0/0', undefined, undefined, key, 10);
rt.addInternetGatewayRoute('testIgwRoute2', undefined, 'pl-1234', undefined, key, 10);
rt.addInternetGatewayRoute('testIgwRoute3', undefined, undefined, '::1', key, 10);
rt.addEgressOnlyIgwRoute('testEigwRoute1', undefined, 'pl-1234', undefined, key, 10);
rt.addEgressOnlyIgwRoute('testEigwRoute2', '0.0.0.0/0', undefined, undefined, key, 10);
rt.addEgressOnlyIgwRoute('testEigwRoute3', undefined, undefined, '::1', key, 10);
rt.addVirtualPrivateGatewayRoute('testVgwRoute', '10.0.30./24', undefined, undefined, key, 10);
rt.addVirtualPrivateGatewayRoute('testVgw2Route', undefined, 'pl-1234', undefined, key, 10);
rt.addVirtualPrivateGatewayRoute('testVgw3Route', undefined, undefined, '::1', key, 10);
rt.addGatewayAssociation('internetGateway');
rt.addGatewayAssociation('virtualPrivateGateway');
/**
 * RouteTable construct test
 */
describe('RouteTable', () => {
  it('addTransitGatewayRoute destinationPrefix list without logRetention throws error', () => {
    function noLogRetention() {
      rt.addTransitGatewayRoute(
        'testRoute2',
        'tgw-1234',
        tgwAttachment,
        undefined,
        'pl-1234',
        undefined,
        key,
        undefined,
      );
    }
    expect(noLogRetention).toThrow(
      new Error('Attempting to add prefix list route without specifying log group retention period'),
    );
  });
  it('addTransitGatewayRoute no destination throws error', () => {
    function noDest() {
      rt.addTransitGatewayRoute('testRoute3', 'tgw-1234', tgwAttachment, undefined, undefined, undefined, key, 10);
    }
    expect(noDest).toThrow(new Error('Attempting to add CIDR route without specifying destination'));
  });
  it('addNatGatewayRoute destinationPrefix list without logRetention throws error', () => {
    function noLogRetention() {
      rt.addNatGatewayRoute(
        'testNgwRoute2',
        ngw.natGatewayId,
        '10.0.3.0/24',
        'destinationPrefixListId',
        undefined,
        key,
        undefined,
      );
    }
    expect(noLogRetention).toThrow(
      new Error('Attempting to add prefix list route without specifying log group retention period'),
    );
  });
  it('addNatGatewayRoute no destination throws error', () => {
    function noDest() {
      rt.addNatGatewayRoute('testNgwRoute3', ngw.natGatewayId, undefined, undefined, undefined, key, 10);
    }
    expect(noDest).toThrow(new Error('Attempting to add CIDR route without specifying destination'));
  });
  it('addInternetGatewayRoute destinationPrefix list without logRetention throws error', () => {
    function noLogRetention() {
      rt.addInternetGatewayRoute('testIgwRoute2', '0.0.0.0/0', 'destinationPrefixListId', undefined, key, undefined);
    }
    expect(noLogRetention).toThrow(
      new Error('Attempting to add prefix list route without specifying log group retention period'),
    );
  });
  it('addInternetGatewayRoute no destination throws error', () => {
    function noDest() {
      rt.addInternetGatewayRoute('testIgwRoute3', undefined, undefined, undefined, key, 10);
    }
    expect(noDest).toThrow(new Error('Attempting to add CIDR route without specifying destination'));
  });

  it('addEgressOnlyIgwRoute destinationPrefix list without logRetention throws error', () => {
    function noLogRetention() {
      rt.addEgressOnlyIgwRoute('testIgwRoute2', '0.0.0.0/0', 'destinationPrefixListId', undefined, key, undefined);
    }
    expect(noLogRetention).toThrow(
      new Error('Attempting to add prefix list route without specifying log group retention period'),
    );
  });
  it('addEgressOnlyIgwRoute no destination throws error', () => {
    function noDest() {
      rt.addEgressOnlyIgwRoute('testIgwRoute3', undefined, undefined, undefined, key, 10);
    }
    expect(noDest).toThrow(new Error('Attempting to add CIDR route without specifying destination'));
  });

  it('addVirtualPrivateGatewayRoute destinationPrefix list without logRetention throws error', () => {
    function noLogRetention() {
      rt.addVirtualPrivateGatewayRoute(
        'testVgwRoute2',
        '0.0.0.0/0',
        'destinationPrefixListId',
        undefined,
        key,
        undefined,
      );
    }
    expect(noLogRetention).toThrow(
      new Error('Attempting to add prefix list route without specifying log group retention period'),
    );
  });
  it('addVirtualPrivateGatewayRoute no destination throws error', () => {
    function noDest() {
      rt.addVirtualPrivateGatewayRoute('testVgwRoute3', undefined, undefined, undefined, key, 10);
    }
    expect(noDest).toThrow(new Error('Attempting to add CIDR route without specifying destination'));
  });
  snapShotTest(testNamePrefix, stack);
});
