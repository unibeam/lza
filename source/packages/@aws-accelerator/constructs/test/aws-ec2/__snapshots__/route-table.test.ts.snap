// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouteTable Construct(RouteTable):  Snapshot Test 1`] = `
{
  "Resources": {
    "CustomPrefixListRouteCustomResourceProviderHandler5B28D077": {
      "DependsOn": [
        "CustomPrefixListRouteCustomResourceProviderRoleD08268B5",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomPrefixListRouteCustomResourceProviderRoleD08268B5",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomPrefixListRouteCustomResourceProviderLogGroup68DB81A5": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsKeyId": {
          "Fn::GetAtt": [
            "testKey1CDDDD5E",
            "Arn",
          ],
        },
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "CustomPrefixListRouteCustomResourceProviderHandler5B28D077",
              },
            ],
          ],
        },
        "RetentionInDays": 10,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "CustomPrefixListRouteCustomResourceProviderRoleD08268B5": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "ec2:CreateRoute",
                    "ec2:DeleteRoute",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                  "Sid": "AllowModifyRoutes",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "RouteTable82FB8FA6": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestRouteTable",
          },
          {
            "Key": "Test-Key",
            "Value": "Test-Value",
          },
        ],
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "RouteTableGatewayAssociation99E18AA3": {
      "DependsOn": [
        "TestVpcInternetGatewayAttachment60E451D5",
      ],
      "Properties": {
        "GatewayId": {
          "Ref": "TestVpcInternetGateway01360C82",
        },
        "RouteTableId": {
          "Ref": "RouteTable82FB8FA6",
        },
      },
      "Type": "AWS::EC2::GatewayRouteTableAssociation",
    },
    "RouteTableVirtualPrivateGatewayAssociation6AD07550": {
      "DependsOn": [
        "TestVpcVirtualPrivateGatewayAttachment5D655F8D",
      ],
      "Properties": {
        "GatewayId": {
          "Ref": "TestVpcVirtualPrivateGateway56B5C340",
        },
        "RouteTableId": {
          "Ref": "RouteTable82FB8FA6",
        },
      },
      "Type": "AWS::EC2::GatewayRouteTableAssociation",
    },
    "RouteTabletest2NgwRoute7F788B8D": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "CustomPrefixListRouteCustomResourceProviderLogGroup68DB81A5",
      ],
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomPrefixListRouteCustomResourceProviderHandler5B28D077",
            "Arn",
          ],
        },
        "routeDefinition": {
          "DestinationPrefixListId": "pl-1234",
          "NatGatewayId": {
            "Ref": "ngwBD1698D5",
          },
          "RouteTableId": {
            "Ref": "RouteTable82FB8FA6",
          },
        },
      },
      "Type": "Custom::PrefixListRoute",
      "UpdateReplacePolicy": "Delete",
    },
    "RouteTabletest3NgwRoute8D00A105": {
      "Properties": {
        "DestinationIpv6CidrBlock": "::1",
        "NatGatewayId": {
          "Ref": "ngwBD1698D5",
        },
        "RouteTableId": {
          "Ref": "RouteTable82FB8FA6",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "RouteTabletestEigwRoute12DFE9140": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "CustomPrefixListRouteCustomResourceProviderLogGroup68DB81A5",
      ],
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomPrefixListRouteCustomResourceProviderHandler5B28D077",
            "Arn",
          ],
        },
        "routeDefinition": {
          "DestinationPrefixListId": "pl-1234",
          "EgressOnlyInternetGatewayId": {
            "Ref": "TestVpcEgressOnlyIgwE0823B1A",
          },
          "RouteTableId": {
            "Ref": "RouteTable82FB8FA6",
          },
        },
      },
      "Type": "Custom::PrefixListRoute",
      "UpdateReplacePolicy": "Delete",
    },
    "RouteTabletestEigwRoute29FFB6E44": {
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "EgressOnlyInternetGatewayId": {
          "Ref": "TestVpcEgressOnlyIgwE0823B1A",
        },
        "RouteTableId": {
          "Ref": "RouteTable82FB8FA6",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "RouteTabletestEigwRoute3C3EC8E23": {
      "Properties": {
        "DestinationIpv6CidrBlock": "::1",
        "EgressOnlyInternetGatewayId": {
          "Ref": "TestVpcEgressOnlyIgwE0823B1A",
        },
        "RouteTableId": {
          "Ref": "RouteTable82FB8FA6",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "RouteTabletestIgwRoute23DD8B83D": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "CustomPrefixListRouteCustomResourceProviderLogGroup68DB81A5",
        "TestVpcInternetGatewayAttachment60E451D5",
      ],
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomPrefixListRouteCustomResourceProviderHandler5B28D077",
            "Arn",
          ],
        },
        "routeDefinition": {
          "DestinationPrefixListId": "pl-1234",
          "GatewayId": {
            "Ref": "TestVpcInternetGateway01360C82",
          },
          "RouteTableId": {
            "Ref": "RouteTable82FB8FA6",
          },
        },
      },
      "Type": "Custom::PrefixListRoute",
      "UpdateReplacePolicy": "Delete",
    },
    "RouteTabletestIgwRoute3141525D4": {
      "DependsOn": [
        "TestVpcInternetGatewayAttachment60E451D5",
      ],
      "Properties": {
        "DestinationIpv6CidrBlock": "::1",
        "GatewayId": {
          "Ref": "TestVpcInternetGateway01360C82",
        },
        "RouteTableId": {
          "Ref": "RouteTable82FB8FA6",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "RouteTabletestIgwRoute3732DB68": {
      "DependsOn": [
        "TestVpcInternetGatewayAttachment60E451D5",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "TestVpcInternetGateway01360C82",
        },
        "RouteTableId": {
          "Ref": "RouteTable82FB8FA6",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "RouteTabletestNgwRouteAE321892": {
      "Properties": {
        "DestinationCidrBlock": "10.0.3.0/24",
        "NatGatewayId": {
          "Ref": "ngwBD1698D5",
        },
        "RouteTableId": {
          "Ref": "RouteTable82FB8FA6",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "RouteTabletestVgw2RouteDD19A5B1": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "CustomPrefixListRouteCustomResourceProviderLogGroup68DB81A5",
        "TestVpcVirtualPrivateGatewayAttachment5D655F8D",
      ],
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomPrefixListRouteCustomResourceProviderHandler5B28D077",
            "Arn",
          ],
        },
        "routeDefinition": {
          "DestinationPrefixListId": "pl-1234",
          "GatewayId": {
            "Ref": "TestVpcVirtualPrivateGateway56B5C340",
          },
          "RouteTableId": {
            "Ref": "RouteTable82FB8FA6",
          },
        },
      },
      "Type": "Custom::PrefixListRoute",
      "UpdateReplacePolicy": "Delete",
    },
    "RouteTabletestVgw3RouteFE177B5B": {
      "DependsOn": [
        "TestVpcVirtualPrivateGatewayAttachment5D655F8D",
      ],
      "Properties": {
        "DestinationIpv6CidrBlock": "::1",
        "GatewayId": {
          "Ref": "TestVpcVirtualPrivateGateway56B5C340",
        },
        "RouteTableId": {
          "Ref": "RouteTable82FB8FA6",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "RouteTabletestVgwRoute717D63D3": {
      "DependsOn": [
        "TestVpcVirtualPrivateGatewayAttachment5D655F8D",
      ],
      "Properties": {
        "DestinationCidrBlock": "10.0.30./24",
        "GatewayId": {
          "Ref": "TestVpcVirtualPrivateGateway56B5C340",
        },
        "RouteTableId": {
          "Ref": "RouteTable82FB8FA6",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "RouteTabletgCEFADE05": {
      "DependsOn": [
        "tgwAttachment1E239374",
      ],
      "Properties": {
        "DestinationCidrBlock": "10.0.5.0/24",
        "RouteTableId": {
          "Ref": "RouteTable82FB8FA6",
        },
        "TransitGatewayId": "tg-1234",
      },
      "Type": "AWS::EC2::Route",
    },
    "RouteTabletgipv684D762E9": {
      "DependsOn": [
        "tgwAttachment1E239374",
      ],
      "Properties": {
        "DestinationIpv6CidrBlock": "::1",
        "RouteTableId": {
          "Ref": "RouteTable82FB8FA6",
        },
        "TransitGatewayId": "tg-1234",
      },
      "Type": "AWS::EC2::Route",
    },
    "RouteTabletgroute63CDF5B7": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "CustomPrefixListRouteCustomResourceProviderLogGroup68DB81A5",
        "tgwAttachment1E239374",
      ],
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomPrefixListRouteCustomResourceProviderHandler5B28D077",
            "Arn",
          ],
        },
        "routeDefinition": {
          "DestinationPrefixListId": "pl-1234",
          "RouteTableId": {
            "Ref": "RouteTable82FB8FA6",
          },
          "TransitGatewayId": "tgw-1234",
        },
      },
      "Type": "Custom::PrefixListRoute",
      "UpdateReplacePolicy": "Delete",
    },
    "TestVpcE77CE678": {
      "Properties": {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": false,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": [
          {
            "Key": "Name",
            "Value": "Test",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "TestVpcEgressOnlyIgwE0823B1A": {
      "Properties": {
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::EgressOnlyInternetGateway",
    },
    "TestVpcInternetGateway01360C82": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "Test",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "TestVpcInternetGatewayAttachment60E451D5": {
      "Properties": {
        "InternetGatewayId": {
          "Ref": "TestVpcInternetGateway01360C82",
        },
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
    "TestVpcVirtualPrivateGateway56B5C340": {
      "Properties": {
        "AmazonSideAsn": 65000,
        "Tags": [
          {
            "Key": "Name",
            "Value": "Test",
          },
        ],
        "Type": "ipsec.1",
      },
      "Type": "AWS::EC2::VPNGateway",
    },
    "TestVpcVirtualPrivateGatewayAttachment5D655F8D": {
      "Properties": {
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
        "VpnGatewayId": {
          "Ref": "TestVpcVirtualPrivateGateway56B5C340",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
    "ngwBD1698D5": {
      "Properties": {
        "AllocationId": {
          "Fn::GetAtt": [
            "ngwEipBD9BC2FE",
            "AllocationId",
          ],
        },
        "SubnetId": {
          "Ref": "testsubnetFBD9E7FF",
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "ngw",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "ngwEipBD9BC2FE": {
      "Properties": {
        "Domain": "vpc",
        "Tags": [
          {
            "Key": "Name",
            "Value": "ngw",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "testKey1CDDDD5E": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "testsubnetFBD9E7FF": {
      "Properties": {
        "AvailabilityZone": "a",
        "CidrBlock": "10.0.2.0/24",
        "Tags": [
          {
            "Key": "Name",
            "Value": "test-subnet",
          },
        ],
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "testsubnetRouteTableAssociationA7927A11": {
      "Properties": {
        "RouteTableId": {
          "Ref": "RouteTable82FB8FA6",
        },
        "SubnetId": {
          "Ref": "testsubnetFBD9E7FF",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "tgwAttachment1E239374": {
      "Properties": {
        "SubnetIds": [
          "subnet-123",
        ],
        "Tags": [
          {
            "Key": "Name",
            "Value": "someTgwAttachment",
          },
        ],
        "TransitGatewayId": "tgw-12324",
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::TransitGatewayAttachment",
    },
  },
}
`;
